### PR TITLE
Use `paragonie/random_compat` for random bytes generation, add CSPRNG-based `Security::generateRandomInt()` and use it everywhere

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | -

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -78,7 +78,8 @@
         "bower-asset/jquery": "3.6.*@stable | 3.5.*@stable | 3.4.*@stable | 3.3.*@stable | 3.2.*@stable | 3.1.*@stable | 2.2.*@stable | 2.1.*@stable | 1.11.*@stable | 1.12.*@stable",
         "bower-asset/inputmask": "~3.2.2 | ~3.3.5",
         "bower-asset/punycode": "1.3.*",
-        "bower-asset/yii2-pjax": "~2.0.1"
+        "bower-asset/yii2-pjax": "~2.0.1",
+        "paragonie/random_compat": "^2.0"
```
Comment (rob006) [inline composer.json:82]: AFAIK constraint should be `>=1`.

https://github.com/paragonie/random_compat#version-conflict-with-other-php-project

**Comment by WinterSilence** [comment]
Maybe require Symphony polyfills 5.6 and later? it would to avoid a lot of crutches (:

File: composer.json (hunk 1 above)
Comment (bizley) [reply composer.json:82]: Let's make it safer and do `^1.1 | ^2.0`.

File: framework/composer.json [hunk 2]
```diff
@@ -73,7 +73,8 @@
         "bower-asset/jquery": "3.6.*@stable | 3.5.*@stable | 3.4.*@stable | 3.3.*@stable | 3.2.*@stable | 3.1.*@stable | 2.2.*@stable | 2.1.*@stable | 1.11.*@stable | 1.12.*@stable",
         "bower-asset/inputmask": "~3.2.2 | ~3.3.5",
         "bower-asset/punycode": "1.3.*",
-        "bower-asset/yii2-pjax": "~2.0.1"
+        "bower-asset/yii2-pjax": "~2.0.1",
+        "paragonie/random_compat": "^2.0"
```
Comment (bizley) [inline framework/composer.json:77]: Same here.

**Review by bizley — APPROVED** [review]

**Comment by lubosdz** [comment]
What is the motivation for adding dependency on another external library? 
I never had issues with built-in randomizers - have there been issues?

From [documentation](https://github.com/paragonie/random_compat):

* `If you are using an unsupported version of PHP, please upgrade as soon as possible.` --- this will not be probably possible for many projects, meaning everybody after upgrading Yii should also upgrade PHP to PHP 7.3+ ??!!
* `If you're on Windows, make sure you enable mcrypt.` --- `mcrypt` is long ago outdated and not supported, replaced with openssl.

Has this additional dependency been somewhere discussed .. ? - perhaps I missed it.

**Comment by samdark** [comment]
@lubosdz PHP 5.4+ is supported. That matches Yii2 supported version range. README means something else, I guess, since PHP 7 has no need for compatibility layer.

> What is the motivation for adding dependency on another external library?

Fixing usage of PRNG in critical places such as CAPTCHA. We could probably implement but that would require too much effort:

1. Implementation of generation of a number in a certain range.
2. Good tests and security review (random_compat was reviewed/created by security experts already).
3. Significant changes in tests for many classes since Yii's service locator should be initialized if we use Security class as a service.

Weighted all that I've decided to go easier way.

**Comment by lubosdz** [comment]
@samdark Thanx for clarifying. Looks like you took personal responsibility for the whole framework :-)

**Comment by WinterSilence** [comment]
@samdark "decisions are made without discussion" - it's bad practice for open source. this approach has killed many projects.
@lubosdz in @samdark defense, I can say that this library is best open source polyfill and required to [Symfony polyfills](https://github.com/symfony/polyfill)

**Pull request merged**